pkg/container/versioned/value.go

```diff
@@ -0,0 +1,440 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package versioned
+
+import (
+	"errors"
+	"fmt"
+	"log/slog"
+	"math"
+	"runtime"
+	"slices"
+	"sync/atomic"
+
+	"github.com/hashicorp/go-hclog"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+type version uint64
+
+const (
+	// invalidVersion is never found from the value range.
+	// Also used as the upper bound for new values, so that the value is
+	// found when looking with 'maxVersion'
+	invalidVersion = version(math.MaxUint64)
+
+	// maxVersion in a VersionHold finds the latest version of all non-removed values
+	maxVersion = version(math.MaxUint64 - 1)
+)
+
+// KeepVersion is an exported version type used when releasing memory held for old versions.
+type KeepVersion version
+
+// NextVersion is an exported version type used to add or remove new versions of values.
+type NextVersion version
+
+var (
+	errInvalidVersion  = errors.New("invalid version")
+	errStaleVersion    = errors.New("stale version")
+	errVersionNotFound = errors.New("version not found")
+)
+
+// VersionHold is used to keep values valid for a specific version from being released, so that they
+// are available for use for as long as the VersionHold is not closed.
+//
+// A special form with a nil coordinator, which is returned by Latest(), always finds the latest
+// versions and does not keep any versions from being released.
+type VersionHold struct {
```

Comment:
So refer to them via an interface instead?

Reply:
So basically other packages need to be able to hold references (pointers) to `VersionHandle`, so some type needs to be exported for that purpose, right? Adding an interface would add functionality that is not needed, so exporting this type, even when it has no exported fields should be fine?

Reply:
Moreover, `VersionHandle` is a struct, and it is deliberately made to have no exported fields. Since it is accessed via a pointer, an exported interface could be used instead. But if it were accessed by value, then that would not be an option.

Reply:
But the proverb goes `Accept interfaces return structs` 🙃 The point is that you're supposed to define interfaces at the site of the consumer, not the producer.

Reply:
Ah - there is one reason which makes it slightly more convenient to have it exported I suppose:

If unexported, and you want to declare functions taking `*versionHold` _outside_ of the `versioned` package you'd have to declare an interface _in that package_ which `*versionHold` implicitly implements, whereas with an exported struct you can just use that. Mostly a preference thing, I suppose - the former has the benefit of the producer being able to swap out the implementation transparently if it conforms to the interface still. 

```go
type version interface {
    Close()
}

func foo(ver version) {} // will happily take *versionHold
```